src/optimize.jl

```diff
@@ -150,42 +151,120 @@ function optimize!(code::CodeInfo, mod::Module)
-                def = quote
-                    function $methname($(argnames...))
-                        return Base.llvmcall($str, $RetType, $ArgType, $(argnames...))
-                    end
-                end
-                f = Core.eval(CompiledCalls, def)
-                stmt.args[1] = QuoteNode(f)
-                deleteat!(stmt.args, 2:4)
+                build_compiled_call!(stmt, methname, Base.llvmcall, stmt.args[2:4], code, idx, nargs, sparams)
                 methodtables[idx] = Compiled()
             end
+        elseif isexpr(stmt, :foreigncall) && isempty(sparams) && scope isa Method
+            f = lookup_stmt(code.code, stmt.args[1])
+            if isa(f, Ptr)
+                f = string(uuid4())
+            elseif isexpr(f, :call)
+                @assert length(f.args) == 3
+                lib = f.args[3] isa String ? f.args[3] : f.args[3].value
+                prefix = f.args[2] isa String ? f.args[2] : f.args[2].value
+                f = Symbol(prefix, '_', lib)
+            end
+            # Punt on non literal ccall arguments for now
+            if !(isa(f, String) || isa(f, Symbol) || isa(f, Ptr))
+                continue
+            end
+            # TODO: Only compile one ccall per call and argument types
+            uuid = uuid4()
+            ustr = replace(string(uuid), '-'=>'_')
+            methname = Symbol("ccall", '_', f, '_', ustr)
+            nargs = stmt.args[5]
+            build_compiled_call!(stmt, methname, :ccall, stmt.args[1:3], code, idx, nargs, sparams)
+            methodtables[idx] = Compiled()
         end
     end
 
     return code, methodtables
 end
+
+# Handle :llvmcall & :foreigncall (issue #28)
+function build_compiled_call!(stmt, methname, fcall, typargs, code, idx, nargs, sparams)
+    argnames = Any[Symbol("arg", string(i)) for i = 1:nargs]
+    # Run a mini-interpreter to extract the types
+    framecode = FrameCode(CompiledCalls, code; optimize=false)
+    frame = Frame(framecode, prepare_framedata(framecode, []))
+    idxstart = idx
+    for i = 2:4
+        idxstart = smallest_ref(code.code, stmt.args[i], idxstart)
+    end
+    frame.pc = idxstart
+    if idxstart < idx
+        while true
+            pc = step_expr!(Compiled(), frame)
+            pc == idx && break
+            pc === nothing && error("this should never happen")
+        end
+    end
+    if fcall == :ccall
+        cfunc, RetType, ArgType = lookup_stmt(code.code, stmt.args[1]), @lookup(frame, stmt.args[2]), @lookup(frame, stmt.args[3])
+        # The result of this is useful to have next to you when reading this code:
+        # f(x, y) =  ccall(:jl_value_ptr, Ptr{Cvoid}, (Float32,Any), x, y)
+        # @code_lowered f(2, 3)
+        args = []
+        for (atype, arg) in zip(ArgType, stmt.args[6:6+nargs-1])
+            if atype === Any
+                push!(args, arg)
+            else
+                @assert arg isa SSAValue
+                unsafe_convert = code.code[arg.id]
```

Comment:
maybe call this `unsafe_convert_expr` or something? Likewise for `cconvert` below.